include/alpaka/stream/StreamCudaRtAsync.hpp

```diff
@@ -238,7 +238,79 @@ namespace alpaka
         namespace traits
         {
             //#############################################################################
-            //! The CUDA RT stream test trait specialization.
+            //! The CUDA RT sync stream enqueue trait specialization.
+            //#############################################################################
+            template<
+                typename TTask>
+            struct Enqueue<
+                stream::StreamCudaRtAsync,
+                TTask>
+            {
+                //#############################################################################
+                //!
+                //#############################################################################
+                struct CallbackSynchronizationData
+                {
+                    std::mutex m_mutex;
+                    std::condition_variable m_event;
+                    bool notified = false;
+                };
+
+                //-----------------------------------------------------------------------------
+                //
+                //-----------------------------------------------------------------------------
+                static void CUDART_CB cudaRtCallback(cudaStream_t /*stream*/, cudaError_t /*status*/, void *arg)
+                {
+                    auto& callbackSynchronizationData = *reinterpret_cast<CallbackSynchronizationData*>(arg);
+
+                    {
+                        std::unique_lock<std::mutex> lock(callbackSynchronizationData.m_mutex);
+                        callbackSynchronizationData.notified = true;
+                    }
+
+                    callbackSynchronizationData.m_event.notify_one();
+                }
+
+                //-----------------------------------------------------------------------------
+                //
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto enqueue(
+                    stream::StreamCudaRtAsync & stream,
+                    TTask const & task)
+                -> void
+                {
+                    auto pCallbackSynchronizationData = std::make_shared<CallbackSynchronizationData>();
+
+                    ALPAKA_CUDA_RT_CHECK(cudaStreamAddCallback(
+                        stream.m_spStreamCudaRtAsyncImpl->m_CudaStream,
+                        cudaRtCallback,
+                        pCallbackSynchronizationData.get(),
+                        0u));
+
+                    std::thread t(
```

Comment:
To create a thread per callback can be very expensive in time. e.g. PIConGPU is spawning over 2000 kernel/memcpy per second and will have over 100 tasks waiting in streams. This means we need to spawn 2k threads/s and have over 100 active threads.

Is it possible to use on thread for all callbacks (waiting in the background), add callbacks to a list and than execute the callback always from the callback thread?

We can move this also to later pull request if it is currently not easy to do.

Reply:
Currently, I would like to merge this as is. I would create a follow-up issue for optimizing this.
The most flexible solution would be a thread pool with a queue of ready callbacks. If there is only one thread, the latency would be highest but it would equal your single thread solution, if there are multiple threads, the latency/resource tradeoff can be adapted per use case.

Reply:
Can we merge this and create the follow up ticket? My upcoming event unit tests are based on some of those stream test helpers.

Reply:
sry I was busy, yes I will merge it